### PR TITLE
Add QSD hotplug vhost-user-blk-pci device testing

### PR DESCRIPTION
**Pull request by qingwangrh**
ID:2154780
Signed-off-by: qingwangrh <qinwang@redhat.com>

### CODE REVIEW

**Comment by qingwangrh** [comment]
@YongxueHong @menli820 @XueqiangWei Please help to review,thanks.

**Comment by qingwangrh** [comment]
python ConfigTest.py --testcase=qsd_hotplug_vubp --iothread_scheme=roundrobin --nr_iothreads=2 --platform=x86_64 --guestname=RHEL.9.2.0,Win2019 --driveformat=virtio_scsi --nicmodel=virtio_net --imageformat=qcow2 --machines=q35,i440fx   --firmware=default_bios --netdst=virbr0 --clone=no --customsparams="cpu_model_flags = 'vmx,+kvm_pv_unhalt'"

 (1/4) Host_RHEL.m9.u2.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.2.0.x86_64.io-github-autotest-qemu.qsd_hotplug_vubp: STARTED
 (1/4) Host_RHEL.m9.u2.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.2.0.x86_64.io-github-autotest-qemu.qsd_hotplug_vubp: PASS (188.34 s)
 (2/4) Host_RHEL.m9.u2.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.2.0.x86_64.io-github-autotest-qemu.qsd_hotplug_vubp.q35: STARTED
 (2/4) Host_RHEL.m9.u2.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.2.0.x86_64.io-github-autotest-qemu.qsd_hotplug_vubp.q35: PASS (188.47 s)
 (3/4) Host_RHEL.m9.u2.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.qsd_hotplug_vubp: STARTED
 (3/4) Host_RHEL.m9.u2.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.qsd_hotplug_vubp: PASS (555.94 s)
 (4/4) Host_RHEL.m9.u2.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.qsd_hotplug_vubp.q35: STARTED
 (4/4) Host_RHEL.m9.u2.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.qsd_hotplug_vubp.q35: PASS (463.74 s)
RESULTS    : PASS 4 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0

File: qemu/tests/cfg/qsd_hotplug_vubp.cfg [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    # Image export attributes
+    qsd_image_export_stg1 = {"type":"vhost-user-blk","num-queues":1}
+    drive_format_stg1 = vhost-user-blk-pci
+    image_vubp_props_stg1=  {"num-queues":1}
+    Linux:
+        guest_cmd = "dd if=/dev/urandom of=%s oflag=direct bs=1M count=1000"
+    Windows:
+        guest_cmd = "WIN_UTILS:\coreutils\DummyCMD.exe %s 1024000000 1"
+    ovmf:
+        no i440fx
```
Comment (menli820) [inline qemu/tests/cfg/qsd_hotplug_vubp.cfg:26]: This seems meaningless.

Comment (qingwangrh) [reply qemu/tests/cfg/qsd_hotplug_vubp.cfg:26]: It avoid to ovmf work with i440fx

Comment (XueqiangWei) [reply qemu/tests/cfg/qsd_hotplug_vubp.cfg:26]: since we only support x86 platform(the condition is in line 6), I think "only q35" is also ok.

Comment (qingwangrh) [reply qemu/tests/cfg/qsd_hotplug_vubp.cfg:26]: it may works on i440fx+seabios

Comment (qingwangrh) [reply qemu/tests/cfg/qsd_hotplug_vubp.cfg:26]: do you mean:
ovmf:
   only q35

**Comment by menli820** [comment]
@qingwangrh   hit a issue when run with win11, add the log to the auto bug, please help to have a check, thanks

File: qemu/tests/cfg/qsd_hotplug_vubp.cfg (hunk 1 above)
Comment (XueqiangWei) [reply qemu/tests/cfg/qsd_hotplug_vubp.cfg:26]: Yes, yours is also OK for me

File: qemu/tests/qsd_hotplug_vubp.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    try:
+        qsd_name = params["qsd_namespaces"]
+        qsd = QsdDaemonDev(qsd_name, params)
+        qsd.start_daemon()
+        img = params["qsd_images_%s" % qsd_name]
+        add_vubp_into_boot(img, params, 6)
+
+        params["start_vm"] = 'yes'
+
+        login_timeout = int(params.get("login_timeout", 360))
```
Comment (menli820) [inline qemu/tests/qsd_hotplug_vubp.py:78]:
```suggestion
        login_timeout = params.get_numeric("login_timeout", 360)
```

Comment (qingwangrh) [reply qemu/tests/qsd_hotplug_vubp.py]: updated

**Comment by menli820** [comment]
LGTM

 (1/2) Host_RHEL.m9.u2.qcow2.virtio_blk.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads.q35: STARTED
 (1/2) Host_RHEL.m9.u2.qcow2.virtio_blk.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads.q35: PASS (1314.12 s)
 (2/2) Host_RHEL.m9.u2.qcow2.virtio_blk.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.qsd_hotplug_vubp.q35: STARTED
 (2/2) Host_RHEL.m9.u2.qcow2.virtio_blk.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.qsd_hotplug_vubp.q35: PASS (324.40 s)
RESULTS    : PASS 2 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0

**Review by YongxueHong — DISMISSED** [review]
LGTM.

File: qemu/tests/qsd_hotplug_vubp.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
+"""QSD hotplug vhost-user-blk-pci device test"""
+import time
+
+from provider.qsd import QsdDaemonDev, add_vubp_into_boot
+from provider.qsd import plug_vubp_devices, unplug_vubp_devices
+from virttest import error_context, env_process, utils_disk, utils_misc
+from virttest.utils_disk import clean_partition_windows
+
+
+@error_context.context_aware
```
Comment (XueqiangWei) [inline qemu/tests/qsd_hotplug_vubp.py:10]: no error_context.context() statement in codes, I think we can remove it. Others LGTM.

 (1/2) Host_RHEL.m9.u2.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.2.0.x86_64.io-github-autotest-qemu.qsd_hotplug_vubp.q35: PASS (145.65 s)
(2/2) Host_RHEL.m9.u2.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win11.x86_64.io-github-autotest-qemu.qsd_hotplug_vubp.q35: PASS (284.71 s)

Comment (qingwangrh) [reply qemu/tests/qsd_hotplug_vubp.py]: updated

**Review by YongxueHong — APPROVED** [review]

**Review by XueqiangWei — APPROVED** [review]

**Pull request merged**